### ISSUE TITLE
docker port command does not update information after docker network disconnect and docker network connect commands

### ISSUE THREAD

**Issue by neotysit** [docker/docker#20918]
Output of `docker version`:

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 16:16:33 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 16:16:33 2016
 OS/Arch:      linux/amd64

```

Output of `docker info`:

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.10.2
Storage Driver: devicemapper
 Pool Name: docker-253:0-50331778-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 2.072 GB
 Data Space Total: 107.4 GB
 Data Space Available: 10.92 GB
 Metadata Space Used: 2.085 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-12-01)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.10.0-327.10.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.797 GiB
Name: docker05.intranet.neotys.com
ID: OTC4:TNC7:PPN7:6KBJ:IBKC:BBOA:AHYS:H4P5:HQAP:AUTO:7CY5:LHVD
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled

```

Provide additional environment details (AWS, VirtualBox, physical, etc.):

Centos 7.2 VM docker host running on ESXi 5.5

List the steps to reproduce the issue:
1. [root@docker05 ~]# docker start ba13a33e1dd2
ba13a33e1dd2

2. [root@docker05 ~]# docker port ba13a33e1dd2
80/tcp -> 0.0.0.0:32777

3. [root@docker05 ~]# curl 127.0.0.1:32777
<html><body><h1>It works!</h1></body></html>

4. [root@docker05 ~]# docker network disconnect bridge ba13a33e1dd2

5. [root@docker05 ~]# curl 127.0.0.1:32777
curl: (7) Failed connect to 127.0.0.1:32777; Connection refused

6. [root@docker05 ~]# docker network connect bridge ba13a33e1dd2

7. [root@docker05 ~]# docker port ba13a33e1dd2
80/tcp -> 0.0.0.0:32777

8. [root@docker05 ~]# curl 127.0.0.1:32777
curl: (7) Failed connect to 127.0.0.1:32777; Connection refused

9. [root@docker05 ~]# curl 127.0.0.1:32778
<html><body><h1>It works!</h1></body></html>

10. [root@docker05 ~]# docker port ba13a33e1dd2
80/tcp -> 0.0.0.0:32777

Describe the results you received:

After "docker network disconnect" and "docker network connect" commands the container port mapping changes and the docker port command should reflect this update

Describe the results you expected:

Despite the port mapping changes the "docker port" command still displays the old port mapping

Provide additional info you think is important:

**Comment by neotysit** [comment]
ERRATA:

Describe the results you received:

Despite the port mapping changes the "docker port" command still displays the old port mapping

Describe the results you expected:

After "docker network disconnect" and "docker network connect" commands the container port mapping changes and the docker port command should reflect this update

Provide additional info you think is important:

**Comment by thaJeztah** [comment]
`docker ps` also seems to be showing the old ports. And, re-connecting to the network (or another network) doesn't seem to publish the ports again / make them accessible.

**Comment by thaJeztah** [comment]
`docker restart <container>` seems to assign new ports and publish them again (after `disconnect` / `connect`). I don't think this is expected, or at least not properly documented, so I labeled this a bug

**Comment by mavenugo** [comment]
we couldn't get https://github.com/docker/libnetwork/pull/810 in for 1.10. I think that is the correct fix to solve this and other port-mapping related issues. We will get it in for 1.11.

**Comment by thaJeztah** [comment]
Thanks @mavenugo, and thanks so much for reporting @neotysit !

**Comment by neotysit** [comment]
;-) thx



*From:* it@neotys.com [mailto:it@neotys.com] *On Behalf Of *Sebastiaan van
Stijn
*Sent:* jeudi 3 mars 2016 17:47
*To:* docker/docker
*Cc:* neotysit
*Subject:* Re: [docker] docker port command does not update information
after docker network disconnect and docker network connect commands (#20918)



Thanks @mavenugo <https://github.com/mavenugo>, and thanks so much for
reporting @neotysit <https://github.com/neotysit> !

—
Reply to this email directly or view it on GitHub
<https://github.com/docker/docker/issues/20918#issuecomment-191848344>.

**Comment by calavera** [comment]
This will be fixed in 1.11. The commit that fixed the issue is in the current vendored version of libnetwork:

https://github.com/docker/libnetwork/commit/c0d12322c1156b0f75de99bb24690561fbde6c32

**Comment by langdead** [comment]
After below chain of operation, exposed port of bridge network won't work. Maybe caused the same reason that the port is updated, but `docker ps -a` doesn't update.

docker run => docker stop => docker start again
```
docker run --name=gogs -p 10022:22 -p 10080:3000 -d gogs/gogs
curl host_ip:10080 #succeeds
docker stop gogs
docker start gogs
curl host_ip:10080 # fails
curl docker_ip:10080 #succeeds
```

After several days' trial, I found `run --net=host *` will work, but with the cons that exporting port is not supported.
